File: ios/Sources/WasmAgentToolsPlugin/WasmAgentToolsPlugin.swift
import Capacitor
import Foundation

private let defaultSandboxFuelLimit: UInt64 = 1_000_000
private let defaultSandboxMaxMemoryBytes: UInt64 = 16 * 1024 * 1024
private let defaultSandboxTimeoutSecs: UInt64 = 30
private let defaultJsFuelLimit: UInt64 = 1_000_000_000
private let defaultJsMaxMemoryBytes: UInt64 = 32 * 1024 * 1024
private let defaultJsTimeoutSecs: UInt64 = 5
private let defaultPythonFuelLimit: UInt64 = 1_000_000_000
private let defaultPythonMaxMemoryBytes: UInt64 = 32 * 1024 * 1024
private let defaultPythonTimeoutSecs: UInt64 = 5

@objc(WasmAgentToolsPlugin)
public class WasmAgentToolsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WasmAgentToolsPlugin"
    public let jsName = "WasmAgentTools"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "createWasmSandbox", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "executeSandbox", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "executeWatSandbox", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createJsRuntime", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "executeJs", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createPythonRuntime", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "executePython", returnType: CAPPluginReturnPromise),
    ]

    private var wasmSandbox: WasmSandbox?
    private var jsRuntime: JsRuntime?
    private var pythonRuntime: PythonRuntime?

    @objc func createWasmSandbox(_ call: CAPPluginCall) {
        do {
            wasmSandbox = nil
            wasmSandbox = try WasmSandbox.create()
            call.resolve()
        } catch {
            call.reject("createWasmSandbox failed: \(error.localizedDescription)")
        }
    }

    @objc func executeSandbox(_ call: CAPPluginCall) {
        guard let wasmSandbox = requireWasmSandbox(call) else { return }
        guard let wasmBase64 = call.getString("wasmBase64") else {
            return call.reject("wasmBase64 is required")
        }
        guard let inputJson = call.getString("inputJson") else {
            return call.reject("inputJson is required")
        }
        guard let wasmBytes = Data(base64Encoded: wasmBase64) else {
            return call.reject("Invalid base64")
        }

        do {
            let result = try wasmSandbox.execute(
                wasmBytes: wasmBytes,
                inputJson: inputJson,
                config: sandboxConfig(from: call.getObject("config"))
            )
            call.resolve(sandboxResultToDictionary(result))
        } catch {
            call.reject("executeSandbox failed: \(error.localizedDescription)")
        }
    }

    @objc func executeWatSandbox(_ call: CAPPluginCall) {
        guard let wasmSandbox = requireWasmSandbox(call) else { return }
        guard let watText = call.getString("watText") else {
            return call.reject("watText is required")
        }
        guard let inputJson = call.getString("inputJson") else {
            return call.reject("inputJson is required")
        }

        do {
            let result = try wasmSandbox.executeWat(
                watText: watText,
                inputJson: inputJson,
                config: sandboxConfig(from: call.getObject("config"))
            )
            call.resolve(sandboxResultToDictionary(result))
        } catch {
            call.reject("executeWatSandbox failed: \(error.localizedDescription)")
        }
    }

    @objc func createJsRuntime(_ call: CAPPluginCall) {
        do {
            jsRuntime = nil
            jsRuntime = try JsRuntime.create()
            call.resolve()
        } catch {
            call.reject("createJsRuntime failed: \(error.localizedDescription)")
        }
    }

    @objc func executeJs(_ call: CAPPluginCall) {
        guard let jsRuntime = requireJsRuntime(call) else { return }
        guard let code = call.getString("code") else {
            return call.reject("code is required")
        }

        do {
            let result = try jsRuntime.execute(
                code: code,
                config: jsConfig(from: call.getObject("config"))
            )
            call.resolve(jsResultToDictionary(result))
        } catch {
            call.reject("executeJs failed: \(error.localizedDescription)")
        }
    }

    @objc func createPythonRuntime(_ call: CAPPluginCall) {
        do {
            pythonRuntime = nil
            pythonRuntime = try PythonRuntime.create()
            call.resolve()
        } catch {
            call.reject("createPythonRuntime failed: \(error.localizedDescription)")
        }
    }

    @objc func executePython(_ call: CAPPluginCall) {
        guard let pythonRuntime = requirePythonRuntime(call) else { return }
        guard let code = call.getString("code") else {
            return call.reject("code is required")
        }

        do {
            let result = try pythonRuntime.execute(
                code: code,
                config: pythonConfig(from: call.getObject("config"))
            )
            call.resolve(pythonResultToDictionary(result))
        } catch {
            call.reject("executePython failed: \(error.localizedDescription)")
        }
    }

    private func requireWasmSandbox(_ call: CAPPluginCall) -> WasmSandbox? {
        guard let wasmSandbox else {
            call.reject("WasmSandbox not initialized — call createWasmSandbox() first")
            return nil
        }
        return wasmSandbox
    }

    private func requireJsRuntime(_ call: CAPPluginCall) -> JsRuntime? {
        guard let jsRuntime else {
            call.reject("JsRuntime not initialized - call createJsRuntime() first")
            return nil
        }
        return jsRuntime
    }

    private func requirePythonRuntime(_ call: CAPPluginCall) -> PythonRuntime? {
        guard let pythonRuntime else {
            call.reject("PythonRuntime not initialized - call createPythonRuntime() first")
            return nil
        }
        return pythonRuntime
    }

    private func uint64(_ value: JSValue?) -> UInt64? {
        if let n = value as? NSNumber { return n.uint64Value }
        return nil
    }

    private func sandboxConfig(from config: JSObject?) -> SandboxConfig {
        let fuelLimit = uint64(config?["fuelLimit"]) ?? defaultSandboxFuelLimit
        let maxMemoryBytes = uint64(config?["maxMemoryBytes"]) ?? defaultSandboxMaxMemoryBytes
        let timeoutSecs = uint64(config?["timeoutSecs"]) ?? defaultSandboxTimeoutSecs
        return SandboxConfig(
            fuelLimit: fuelLimit,
            maxMemoryBytes: maxMemoryBytes,
            timeoutSecs: timeoutSecs
        )
    }

    private func jsConfig(from config: JSObject?) -> JsConfig {
        let fuelLimit = uint64(config?["fuelLimit"]) ?? defaultJsFuelLimit
        let maxMemoryBytes = uint64(config?["maxMemoryBytes"]) ?? defaultJsMaxMemoryBytes
        let timeoutSecs = uint64(config?["timeoutSecs"]) ?? defaultJsTimeoutSecs
        return JsConfig(
            timeoutSecs: timeoutSecs,
            fuelLimit: fuelLimit,
            maxMemoryBytes: maxMemoryBytes
        )
    }

    private func pythonConfig(from config: JSObject?) -> PythonConfig {
        let fuelLimit = uint64(config?["fuelLimit"]) ?? defaultPythonFuelLimit
        let maxMemoryBytes = uint64(config?["maxMemoryBytes"]) ?? defaultPythonMaxMemoryBytes
        let timeoutSecs = uint64(config?["timeoutSecs"]) ?? defaultPythonTimeoutSecs
        return PythonConfig(
            timeoutSecs: timeoutSecs,
            fuelLimit: fuelLimit,
            maxMemoryBytes: maxMemoryBytes
        )
    }

    private func sandboxResultToDictionary(_ result: SandboxResult) -> [String: Any] {
        [
            "output": result.output,
            "fuelConsumed": result.fuelConsumed,
            "durationMs": result.durationMs,
        ]
    }

    private func jsResultToDictionary(_ result: JsResult) -> [String: Any] {
        var payload: [String: Any] = [
            "output": result.output,
            "fuelConsumed": result.fuelConsumed,
            "durationMs": result.durationMs,
        ]
        if let error = result.error {
            payload["error"] = error
        }
        return payload
    }

    private func pythonResultToDictionary(_ result: PythonResult) -> [String: Any] {
        var payload: [String: Any] = [
            "output": result.output,
            "fuelConsumed": result.fuelConsumed,
            "durationMs": result.durationMs,
        ]
        if let error = result.error {
            payload["error"] = error
        }
        return payload
    }
}

File: android/src/main/java/com/t6x/plugins/wasmagenttools/WasmAgentToolsPlugin.kt
package com.t6x.plugins.wasmagenttools

import android.util.Base64
import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import uniffi.wasm_agent_tools_ffi.JsConfig
import uniffi.wasm_agent_tools_ffi.JsResult
import uniffi.wasm_agent_tools_ffi.JsRuntime
import uniffi.wasm_agent_tools_ffi.PythonConfig
import uniffi.wasm_agent_tools_ffi.PythonResult
import uniffi.wasm_agent_tools_ffi.PythonRuntime
import uniffi.wasm_agent_tools_ffi.SandboxConfig
import uniffi.wasm_agent_tools_ffi.SandboxResult
import uniffi.wasm_agent_tools_ffi.WasmSandbox

private const val DEFAULT_FUEL_LIMIT = 1_000_000UL
private const val DEFAULT_MAX_MEMORY_BYTES = 16_777_216UL
private const val DEFAULT_TIMEOUT_SECS = 30UL
private const val DEFAULT_JS_FUEL_LIMIT = 1_000_000_000UL
private const val DEFAULT_JS_MAX_MEMORY_BYTES = 33_554_432UL
private const val DEFAULT_JS_TIMEOUT_SECS = 5UL
private const val DEFAULT_PYTHON_FUEL_LIMIT = 1_000_000_000UL
private const val DEFAULT_PYTHON_MAX_MEMORY_BYTES = 33_554_432UL
private const val DEFAULT_PYTHON_TIMEOUT_SECS = 5UL

@CapacitorPlugin(name = "WasmAgentTools")
class WasmAgentToolsPlugin : Plugin() {
    companion object { private const val TAG = "WasmAgentTools" }
    private var wasmSandbox: WasmSandbox? = null
    private var jsRuntime: JsRuntime? = null
    private var pythonRuntime: PythonRuntime? = null

    // Use dedicated threads instead of bridge.execute — the Capacitor bridge
    // handler is a serial HandlerThread, and @PluginMethod already runs on it.
    // Posting back via bridge.execute deadlocks because the handler is blocked
    // waiting for the posted Runnable which is queued behind the current call.

    @PluginMethod
    fun createWasmSandbox(call: PluginCall) {
        call.setKeepAlive(true)
        Thread {
            try {
                wasmSandbox?.close()
                wasmSandbox = WasmSandbox.create()
                call.resolve()
            } catch (e: Exception) {
                call.reject("createWasmSandbox failed: ${e.message}", e)
            }
        }.start()
    }

    @PluginMethod
    fun executeSandbox(call: PluginCall) {
        val sandbox = requireWasmSandbox(call) ?: return
        val wasmBase64 = call.getString("wasmBase64") ?: return call.reject("wasmBase64 is required")
        val inputJson = call.getString("inputJson") ?: return call.reject("inputJson is required")

        call.setKeepAlive(true)
        Thread {
            try {
                val wasmBytes = Base64.decode(wasmBase64, Base64.DEFAULT)
                val config = toSandboxConfig(call.getObject("config"))
                val result = sandbox.`execute`(wasmBytes, inputJson, config)
                call.resolve(sandboxResultToJs(result))
            } catch (e: Exception) {
                call.reject("executeSandbox failed: ${e.message}", e)
            }
        }.start()
    }

    @PluginMethod
    fun executeWatSandbox(call: PluginCall) {
        val sandbox = requireWasmSandbox(call) ?: return
        val watText = call.getString("watText") ?: return call.reject("watText is required")
        val inputJson = call.getString("inputJson") ?: return call.reject("inputJson is required")

        call.setKeepAlive(true)
        Thread {
            try {
                val config = toSandboxConfig(call.getObject("config"))
                val result = sandbox.`executeWat`(watText, inputJson, config)
                call.resolve(sandboxResultToJs(result))
            } catch (e: Exception) {
                call.reject("executeWatSandbox failed: ${e.message}", e)
            }
        }.start()
    }

    @PluginMethod
    fun createJsRuntime(call: PluginCall) {
        call.setKeepAlive(true)
        Thread {
            try {
                jsRuntime?.close()
                jsRuntime = JsRuntime.create()
                call.resolve()
            } catch (e: Exception) {
                call.reject("createJsRuntime failed: ${e.message}", e)
            }
        }.start()
    }

    @PluginMethod
    fun executeJs(call: PluginCall) {
        val runtime = requireJsRuntime(call) ?: return
        val code = call.getString("code") ?: return call.reject("code is required")

        call.setKeepAlive(true)
        Thread {
            try {
                val config = toJsConfig(call.getObject("config"))
                val result = runtime.`execute`(code, config)
                call.resolve(jsResultToJs(result))
            } catch (e: Exception) {
                call.reject("executeJs failed: ${e.message}", e)
            }
        }.start()
    }

    @PluginMethod
    fun createPythonRuntime(call: PluginCall) {
        Log.i(TAG, "createPythonRuntime: dispatching to Thread (caller=${Thread.currentThread().name})")
        call.setKeepAlive(true)
        Thread {
            Log.i(TAG, "createPythonRuntime: Thread started (${Thread.currentThread().name})")
            try {
                pythonRuntime?.close()
                Log.i(TAG, "createPythonRuntime: calling PythonRuntime.create()...")
                val start = System.currentTimeMillis()
                pythonRuntime = PythonRuntime.create()
                Log.i(TAG, "createPythonRuntime: done in ${System.currentTimeMillis() - start}ms")
                call.resolve()
            } catch (e: Exception) {
                Log.e(TAG, "createPythonRuntime: FAILED: ${e.message}", e)
                call.reject("createPythonRuntime failed: ${e.message}", e)
            }
        }.start()
    }

    @PluginMethod
    fun executePython(call: PluginCall) {
        val runtime = requirePythonRuntime(call) ?: return
        val code = call.getString("code") ?: return call.reject("code is required")

        call.setKeepAlive(true)
        Thread {
            try {
                val config = toPythonConfig(call.getObject("config"))
                val result = runtime.`execute`(code, config)
                call.resolve(pythonResultToJs(result))
            } catch (e: Exception) {
                call.reject("executePython failed: ${e.message}", e)
            }
        }.start()
    }

    private fun requireWasmSandbox(call: PluginCall): WasmSandbox? {
        val sandbox = wasmSandbox
        if (sandbox == null) {
            call.reject("WasmSandbox not initialized — call createWasmSandbox() first")
        }
        return sandbox
    }

    private fun requireJsRuntime(call: PluginCall): JsRuntime? {
        val runtime = jsRuntime
        if (runtime == null) {
            call.reject("JsRuntime not initialized - call createJsRuntime() first")
        }
        return runtime
    }

    private fun requirePythonRuntime(call: PluginCall): PythonRuntime? {
        val runtime = pythonRuntime
        if (runtime == null) {
            call.reject("PythonRuntime not initialized - call createPythonRuntime() first")
        }
        return runtime
    }

    private fun toSandboxConfig(config: JSObject?): SandboxConfig {
        val fuelLimit = config?.optLong("fuelLimit", DEFAULT_FUEL_LIMIT.toLong())?.toULong()
            ?: DEFAULT_FUEL_LIMIT
        val maxMemoryBytes =
            config?.optLong("maxMemoryBytes", DEFAULT_MAX_MEMORY_BYTES.toLong())?.toULong()
                ?: DEFAULT_MAX_MEMORY_BYTES
        val timeoutSecs = config?.optLong("timeoutSecs", DEFAULT_TIMEOUT_SECS.toLong())?.toULong()
            ?: DEFAULT_TIMEOUT_SECS
        return SandboxConfig(
            fuelLimit = fuelLimit,
            maxMemoryBytes = maxMemoryBytes,
            timeoutSecs = timeoutSecs,
        )
    }

    private fun toJsConfig(config: JSObject?): JsConfig {
        val fuelLimit =
            config?.optLong("fuelLimit", DEFAULT_JS_FUEL_LIMIT.toLong())?.toULong()
                ?: DEFAULT_JS_FUEL_LIMIT
        val maxMemoryBytes =
            config?.optLong("maxMemoryBytes", DEFAULT_JS_MAX_MEMORY_BYTES.toLong())?.toULong()
                ?: DEFAULT_JS_MAX_MEMORY_BYTES
        val timeoutSecs =
            config?.optLong("timeoutSecs", DEFAULT_JS_TIMEOUT_SECS.toLong())?.toULong()
                ?: DEFAULT_JS_TIMEOUT_SECS
        return JsConfig(
            timeoutSecs = timeoutSecs,
            fuelLimit = fuelLimit,
            maxMemoryBytes = maxMemoryBytes,
        )
    }

    private fun toPythonConfig(config: JSObject?): PythonConfig {
        val fuelLimit =
            config?.optLong("fuelLimit", DEFAULT_PYTHON_FUEL_LIMIT.toLong())?.toULong()
                ?: DEFAULT_PYTHON_FUEL_LIMIT
        val maxMemoryBytes =
            config?.optLong("maxMemoryBytes", DEFAULT_PYTHON_MAX_MEMORY_BYTES.toLong())?.toULong()
                ?: DEFAULT_PYTHON_MAX_MEMORY_BYTES
        val timeoutSecs =
            config?.optLong("timeoutSecs", DEFAULT_PYTHON_TIMEOUT_SECS.toLong())?.toULong()
                ?: DEFAULT_PYTHON_TIMEOUT_SECS
        return PythonConfig(
            timeoutSecs = timeoutSecs,
            fuelLimit = fuelLimit,
            maxMemoryBytes = maxMemoryBytes,
        )
    }

    private fun sandboxResultToJs(result: SandboxResult): JSObject {
        val payload = JSObject()
        payload.put("output", result.output)
        payload.put("fuelConsumed", result.fuelConsumed.toLong())
        payload.put("durationMs", result.durationMs.toLong())
        return payload
    }

    private fun jsResultToJs(result: JsResult): JSObject {
        val payload = JSObject()
        payload.put("output", result.output)
        if (result.error != null) {
            payload.put("error", result.error)
        }
        payload.put("fuelConsumed", result.fuelConsumed.toLong())
        payload.put("durationMs", result.durationMs.toLong())
        return payload
    }

    private fun pythonResultToJs(result: PythonResult): JSObject {
        val payload = JSObject()
        payload.put("output", result.output)
        if (result.error != null) {
            payload.put("error", result.error)
        }
        payload.put("fuelConsumed", result.fuelConsumed.toLong())
        payload.put("durationMs", result.durationMs.toLong())
        return payload
    }
}
